### ISSUE TITLE
getting error java.lang.ClassCastException: cannot assign instance of java.lang.invoke.SerializedLambda to field org.apache.spark.sql.catalyst.expressions.ScalaUDF.f of type scala.Function1 in instance of org.apache.spark.sql.catalyst.expressions.ScalaUDF

### ISSUE THREAD

**Issue by cometta** [exacaster/lighter#142]
my driver and executors pod are using the same docker image apache/spark-py:v3.3.0" , lighter server is using ghcr.io/exacaster/lighter:0.0.38-spark3.3.0 , but i keep getting scala error, any idea how to troubleshoot? Does it related to I did not set any value for LIGHTER_SPARK_HISTORY_SERVER_URL ?

```
22/09/30 03:12:01 INFO TaskSchedulerImpl: Adding task set 0.0 with 2 tasks resource profile 0
22/09/30 03:12:01 INFO TaskSetManager: Starting task 0.0 in stage 0.0 (TID 0) (10.42.4.63, executor 1, partition 0, PROCESS_LOCAL, 5169 bytes) taskResourceAssignments Map()
22/09/30 03:12:02 INFO BlockManagerInfo: Added broadcast_2_piece0 in memory on 10.42.4.63:40431 (size: 35.1 KiB, free: 413.9 MiB)
22/09/30 03:12:02 INFO TaskSetManager: Starting task 1.0 in stage 0.0 (TID 1) (10.42.4.63, executor 1, partition 1, PROCESS_LOCAL, 5393 bytes) taskResourceAssignments Map()
22/09/30 03:12:02 WARN TaskSetManager: Lost task 0.0 in stage 0.0 (TID 0) (10.42.4.63 executor 1): java.lang.ClassCastException: cannot assign instance of java.lang.invoke.SerializedLambda to field org.apache.spark.sql.catalyst.expressions.ScalaUDF.f of type scala.Function1 in instance of org.apache.spark.sql.catalyst.expressions.ScalaUDF
	at java.base/java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(Unknown Source)
	at java.base/java.io.ObjectStreamClass$FieldReflector.checkObjectFieldValueTypes(Unknown Source)
	at java.base/java.io.ObjectStreamClass.checkObjFieldValueTypes(Unknown Source)
	at java.base/java.io.ObjectInputStream.defaultCheckFieldValues(Unknown Source)
	at java.base/java.io.ObjectInputStream.readSerialData(Unknown Source)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(Unknown Source)
	at java.base/java.io.ObjectInputStream.readObject0(Unknown Source)
	at java.base/java.io.ObjectInputStream.defaultReadFields(Unknown Source)
	at
```

**Comment by Minutis** [comment]
`LIGHTER_SPARK_HISTORY_SERVER_URL` is an optional parameter, so I doubt that this is the problem. The log snippet you gave indicates that this stacktrace is WARN level and it should not affect your jobs. Could you please provide more information when are you getting this error and what happens then? Also some more information on the configuration you are using would also help.

**Comment by pdambrauskas** [comment]
Strange error, we are using `delta`on our lighter deployment with no problems. Can you provide us more detail on your notebook session configuration:

- which version of Lighter are you running
- what docker image are you using (if it is not a public image, specify Spark version and Delta version)
- value of `LIGHTER_SESSION_DEFAULT_CONF`
- do you provide any additional session configuration through `%%configure` macro?

**Comment by pdambrauskas** [comment]
Thanks for details, we use custom Docker images with all needed jar inside, so we did not face problems related to `spark.jars.packages`. Currently we are testing your use case with `apache/spark-py:v3.3.0` image and `spark.jars.packages` configuration flag, we'll  see if we can find what is the problem.

**Comment by pdambrauskas** [comment]
We do not use `%%sql` macros for our use cases, so cannot comment much on that.
Lighter does not handle any conversions for %%sql, %pretty, etc, so all the conversions come from sparkmagic.

**Comment by pdambrauskas** [comment]
This is the place, where `take` is added:
https://github.com/jupyter-incubator/sparkmagic/blob/2e4c2a2aa153e32a6428de863bd829ec03844609/sparkmagic/sparkmagic/livyclientlib/sqlquery.py#L132

**Comment by pdambrauskas** [comment]
This is part of Spark functionality, Lighter has nothing to do with it.

My guess is - Spark ignores your provided metastor, because by default hive support is not enabled on Spark Session applications, created by lighter. 

You can enable Hive support by setting `spark.sql.catalogImplementation` to `hive`. Maybe that would help with your problem.

**Comment by Minutis** [comment]
> an update, after add this line in LIGHTER_SESSION_DEFAULT_CONF
> 
> ```
>  "spark.jars": "https://repo1.maven.org/maven2/io/delta/delta-core_2.12/2.1.0/delta-core_2.12-2.1.0.jar
> ```
> 
> df = spark.read.format("delta").load("s3a://file/path/deltadataset") is working. do you know why can't just add to `"spark.jars.packages": "io.delta:delta-core_2.12:2.1.0",` ? i did it for other spark apps. is it the same as spark-submit --packages? i can't find any option to add --packages
> 
> now back to the %%sql , when execute with 'show databases' or ' select * from delta.`s3a://bucket/delta/files` ', i get error
> 
> ```
> An error was encountered:
> AttributeError
> [Traceback (most recent call last):
> ,   File "/tmp/spark-71478cb7-605b-457e-acd0-9cd5708b8b53/shell_wrapper.py", line 112, in exec
>     self._exec_then_eval(code.rstrip())
> ,   File "/tmp/spark-71478cb7-605b-457e-acd0-9cd5708b8b53/shell_wrapper.py", line 106, in _exec_then_eval
>     exec(compile(last, '<string>', 'single'), self.globals)
> ,   File "<string>", line 1, in <module>
> , AttributeError: 'function' object has no attribute 'take'
> ]
> ```
> 
> did you try use %%sql to direct query from delta.`s3....` ?

Took a while to test everything out. But it seems that at the moment packages fetched by `--packages` (`spark.jars.packages`) are not available in drivers because of this bug https://issues.apache.org/jira/browse/SPARK-35084. `spark.jars` is a different story: jars added there are shared across the executors. Unfortunately, in `spark.jars` you have to specify all jars including transient dependencies and it's not really that convenient.

I couldn't find any easy way to achieve similar functionality as `--packages` (`spark.jars.packages`). You should probably somehow share the jars in shared storage across all pods and use `spark.jars` with globs i.e. `spark.jars`: `/<path>/<to>/<shared>/<storage>/*`.

**Comment by Minutis** [comment]
Not sure why https://github.com/apache/spark/pull/32397 is closed. But maybe we will see this bug fixed in the future.

**Comment by Minutis** [comment]
> @Minutis sorry for the confusion, i able to use --packages when doing spark-submit in k8s for other app. do you mean --packages is not yet support when using spark python library in lighter? lighter is using python code to do spark-submit to trigger driver pod right?

Lighter is using Java API to submit jobs. `--packages` only pull the jars to the driver pod. If you are using any of the classes from a pulled JAR in executors it should give you an error. In the applications that work maybe you are using new JARs only in the driver? Otherwise I am not sure how you got it working - I couldn't :)

**Comment by Minutis** [comment]
> 

And you are doing this on Kubernetes cluster?

**Comment by Minutis** [comment]
> yes.

Using Spark 3?

**Comment by Minutis** [comment]
Ok I am confused :( Let me put together the whole situation so that you can confirm if I understand it correctly. You have two jobs:

1. **Not working.** You use Lighter with Notebook. Executing anything related to Delta is failing with the error you wrote in the first message. Running on K8s with 3.3.0 spark both Lighter and for spark.kubernetes.image?
2. **Working,** You use manual spark-submit command using 3.3.0 for submit and for spark.kubernetes.image and the job is executing something related to Delta successfully?

If I am correct, could you please provide the full spark-submit command?

**Comment by Minutis** [comment]
Hey, sorry for the late reply. As I understand we can close the issue since you no longer have any problems. If you are still having issues - let us know.

In any case, https://issues.apache.org/jira/browse/SPARK-35084 was fixed with 3.4 Spark. And we have packaged 3.4 Spark since [0.0.45-spark3.4.0](https://github.com/orgs/exacaster/packages/container/lighter/85477373?tag=0.0.45-spark3.4.0) version. That means you can already use `spark.jars.packages` instead of `spark.jars`.

P.S. Of course check the latest version with 3.4 Spark bundled which is [0.0.47-spark3.4.0](https://github.com/orgs/exacaster/packages/container/lighter/92163127?tag=0.0.47-spark3.4.0) at the moment I am writing this.
